cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/MacroParser.java

```diff
@@ -139,11 +160,24 @@ private MacroMetadata getMacroFunctionMetadata(int startIndex, int endIndex, Str
     String macroFunction = replaceEscapedSyntax(macroStr.substring(0, argsStartIndex));
 
     String[] args = Iterables.toArray(Splitter.on(ARGUMENT_DELIMITER).split(arguments), String.class);
-    return new MacroMetadata(macroEvaluator.evaluate(macroFunction, args), startIndex, endIndex);
+    // if function evaluation is disabled, or this is not a whitelisted function, don't perform any evaluation.
+    if (functionsEnabled && (functionWhitelist.isEmpty() || functionWhitelist.contains(macroFunction))) {
```

Comment:
it will be better to have this logic in a private method `shouldEvaluateMacroFunction` also its a little non intuitive that when whileList is empty it will be evaluated by default. will be good to clarify.

Reply:
I don't think it's better to have a separate method because it would be a single line and it would only be called here. It would be another level of indirection for no benefit.

I can add a comment about the empty whitelist.